Stochastic Gradient Descent

junseo-park opened the pull request:
Covered topics:
- Limitations of Gradient Descent
- Stochastic Gradient Descent
- Visualizing Stochastic Gradient Descent
- Mini-batch Gradient Descent
- Selecting batch size
- Summary

junseo-park commented:
I was wondering if we should add these topics:
- Manually implementing SGD and Mini-batch GD in Python, like we do in [11.2 for Batch GD](https://www.textbook.ds100.org/ch11/gradient_descent_define.html#Defining-the-minimize-Function)
- Using [SGDClassifier](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier) for SGD ([`.fit()`](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier.fit)) and Mini-batch GD ([`.partial_fit()`](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html#sklearn.linear_model.SGDClassifier.partial_fit))
- Example using `lebron.csv`

I'm not sure if these are entirely necessary, since the class as it stands only touches upon the concept of SGD/Mini-batch GD and doesn't ever use them in homework/projects; in any case, they'd be useful information. Thoughts?

SamLau95 commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+
+## Selecting the Mini-Batch Size
+
+The mini-batch size is often set to small numbers such as 1 or 2 to minimize computational cost. However, when working with parallel machines such as a GPU or a distributed computing system, we can increase the mini-batch size to take advantage of the hardware. Since mini-batches can be computed in parallel, the runtime of larger mini-batches is roughly equivalent to the runtime of stochastic gra ...
```
I *think* mini-batch sizes are often set to somewhere around 10-100 depending on the GPU memory. I don't think we need to mention the case where you run mini-batch without a GPU since mini-batch isn't super useful without one. Instead, we should mention how a relatively small batch size increases the accuracy of the gradient without increasing computation time when using a GPU. (For reference: https://stats.stackexchange.com/questions/164876/tradeoff-batch-size-vs-number-of-iterations-to-train-a-neural-network .)

Also, we should explain the GPU abbreviation before using it.

junseo-park replied:
Semi-fixed; is the stackexchange post in the scope of the class?

SamLau95 replied:
I don't think the stackexchange post is in scope; I just included it for your reference. I think we should just say the batches are often 10-100.

SamLau95 commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+
+# Summary
+
+Batch gradient descent allows us to minimize the loss of the cost function at each iteration and take direct steps towards the loss-minimizing $\theta$ parameters. However, using the entire batch of $n$ observations can become computationally expensive for large $n$; in such scenarios, we use stochastic gradient descent or mini-batch gradient descent. While they may take noisier paths to the loss ...
```
nit: GD doesn't minimize the loss "at each iteration"; it moves \theta in the direction that lowers cost. "in such scenarios" is bit wordy. In the last sentence, "they" is a bit ambiguous too. We don't really say that paths are "noisier" than others. How about:

We use gradient descent to iteratively improve model parameters until the model achieves low cost. Since batch gradient descent is computationally intractable with large datasets, we often use stochastic gradient descent to fit models instead. With a GPU, mini-batch gradient descent can converge more quickly than stochastic gradient descent.

Although stochastic gradient descent typically takes more individual updates to converge than batch gradient descent, stochastic gradient descent often converges more quickly because it computes updates faster.

junseo-park replied:
Fixed

SamLau95 commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+
+Batch gradient descent allows us to minimize the loss of the cost function at each iteration and take direct steps towards the loss-minimizing $\theta$ parameters. However, using the entire batch of $n$ observations can become computationally expensive for large $n$; in such scenarios, we use stochastic gradient descent or mini-batch gradient descent. While they may take noisier paths to the loss ...
+
+While we have made the distinction between stochastic and mini-batch gradient descent, stochastic gradient descent is sometimes used as an umbrella term that encompasses the selection of a mini-batch of any size. 
```
I think the belongs in the mini-batch gradient section, not the section summary.

junseo-park replied:
Fixed

SamLau95 commented:
As you can probably tell, we're being very strict about phrasing and wording now since we're hoping that this textbook will become an important resource for the class. I know that there are a lot of corrections but I hope the detailed feedback will improve your writing.

Let me know if you disagree with any of the comments and I'd be happy to talk it out since it'll probably help both of us write better :) .

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+\nabla_\theta L(\theta, y) \approx -\frac{1}{|\mathcal{B}|} \sum_{i\in\mathcal{B}}(y_i - \sigma_i)X_i
+$$
+
+In order to select our mini-batch, we could randomly sample at each iteration. In practice, we randomly shuffle the training data and select our mini-batch by looping through the shuffled data. This also applies to stochastic gradient descent.
```
I think including epochs is a good idea, but I didn't include it on my previous draft to be consistent with SP18 Lecture 22, when Joey said "They get to the end, and they probably should shuffle their data, but normally they go back to the beginning". 

Epochs weren't really ever covered in the class, but if the textbook has more leeway to go in-depth, I'm all for adding it! I've added it in to the next draft and we can discuss further?

SamLau95 replied:
Ah I see. Thanks for being consistent with lecture! Let's talk about epochs and this topic tonight.

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+
+# Mini-batch Gradient Descent
+
+We can improve our estimates by increasing the number of observations that we select at each iteration. In **mini-batch gradient descent**, we take a simple random sample (known as a mini-batch) of observations and use the average of their loss gradients to construct an estimate of the true gradient of the cross entropy cost. Since our sample is randomly selected, the expectation of this estimate ...
```
Fixed

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+
+By definition, the gradient of the cross entropy cost is the average (taken over the entire batch of $n$ observations) of each observation's loss gradient, $-(y_i - \sigma_i)X_i$. This is then computed at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve.
+
+# Stochastic Gradient Descent
```
Fixed

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+\theta_{t+1} = \theta_t - \alpha \cdot \nabla_\theta L(\theta, y)
+$$
+
+In logistic regression, we use the cross entropy cost as our loss function. Then we can replace  $\nabla_\theta L(\theta, y)$ with the gradient of the cross entropy cost to find the gradient descent algorithm specific to logistic regression. Letting $ \sigma_i = f_\hat{\theta}(X_i) = \sigma(X_i \cdot \hat \theta) $, this becomes:
```
Fixed

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+\nabla_\theta L(\theta, y) \approx -(y_i - \sigma_i)X_i
+$$
+
+We can then plug this back into the general formula for gradient descent to find estimates for $\theta$:
```
Fixed

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+![](gd.png)
+![](sgd.png)
+
+At each iteration of batch gradient descent, we move in the direction of the true gradient of the loss (shown by the ellipses). On the other hand, at each iteration of stochastic gradient, our steps may not lead us in the direction of the true gradient; however, our $\theta$ estimates eventually converge to the true loss-minimizing values. At large values of $n$, this is much more desirable due t ...
```
Fixed

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+
+# Limitations of Gradient Descent
+
+In Chapter 11, we discussed gradient descent (also known as **batch gradient descent**), which is used to converge at the $\theta$ parameters that minimize a loss function $L(\theta, y)$.
```
Fixed

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+\theta_{t+1} = \theta_t - \alpha \cdot \left(- \frac{1}{n} \sum_{i=1}^{n} \left(y_i - \sigma_i\right) X_i \right)
+$$
+
+By definition, the gradient of the cross entropy cost is the average (taken over the entire batch of $n$ observations) of each observation's loss gradient, $-(y_i - \sigma_i)X_i$. This is then computed at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve.
```
Fixed

junseo-park commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,83 @@
+
+# Stochastic Gradient Descent
+
+Our solution is **stochastic gradient descent**, a technique in which we approximate the true gradient of the cross entropy cost by using just one randomly chosen observation. Since the observation is chosen randomly, we expect that this observation's loss gradient is equal to the true gradient of the cross entropy cost. Our approximation of the gradient of the cross entropy cost is shown below,  ...
```
Fixed

junseo-park commented:
I thought about moving this to the Gradient Descent chapter, but in the class we only cover it once we get to classification and logistic regression. I do think it makes more sense in Ch11, but the lecture examples use cross entropy cost.

SamLau95 commented:
Ah I see. Okay, let's leave it in ch16 for now and see how it feels.

junseo-park commented:
Not yet ready for merge; more fixes required in `notebooks/ch16/classification_sgd.ipynb`.

In the "Move SGD to ch11" commit:
- Moved the theory of SGD into ch11 (`notebooks/ch11/gradient_stochastic.md`)
  - Edited equations to use the cost functions in scope by ch11
  - Manual implementations of SGD and Mini-batch GD in Python
  - Found that ch11 example (modeling tip percent with a constant value) is too simple of a model for SGD/mini-batch to converge
- Replaced previous ch16 SGD with a modified version (`notebooks/ch16/classification_sgd.md`) that incorporates cross entropy loss into general formulas
  - Attempted to come up with examples; however, this was a challenge - I don't think `lebron.csv` was enough observations for SGD/mini-batch to showcase their strengths. 
  - TODO: What datasets could we use for an example?

ananthagarwal commented on notebooks/ch11/gradient_stochastic.ipynb:
```diff
@@ -0,0 +1,285 @@
+    "But for more complicated models and cost functions, there may not be simple algebraic expressions that yield the cost-minimizing $\\hat\\theta$ values. Our solution was a new method called gradient descent, which iteratively updates $\\hat\\theta$ until it doesn't change between iterations. To complete an iteration of gradient descent, we calculate the following:\n",
+    "\n",
+    "$$\n",
+    "\\hat\\theta_{t+1} = \\hat\\theta_t - \\alpha \\cdot \\nabla_{\\hat\\theta} L(\\hat\\theta, y)\n",
```
nit: I think we should specify that the gradient is evaluated at theta = theta_t, since currently the equation states to take the gradient but it should be clear that it is the gradient at theta_t (this is also something I think we should clarify in the other Chapter 11 sections).

Also one thing that could get confusing is the difference between theta and theta_hat, since 11.2 uses theta but we have theta_hat here. Let's keep theta_hat as the optimal theta determined at the end of gradient descent, so I think we should relabel the theta_hats here.

junseo-park replied:
@SamLau95 talked about this previously in another comment; I think he intended to switch the thetas in 11.2 to theta_hats? https://github.com/DS-100/textbook/pull/35#discussion_r196255249

ananthagarwal commented on notebooks/ch11/gradient_stochastic.ipynb:
```diff
@@ -0,0 +1,285 @@
+    "\n",
+    "\n",
+    "\n",
+    "This process is then repeated at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve."
```
Referring to $ t $ here is a little confusing since it isn't used in the above derivation, so perhaps this can be written without the $ t $: "... at each iteration of the gradient ..."

Instead when you mention the general equation involving theta_t+1 and theta_t in the above section, you could clarify that theta_t is the current choice/value of theta and theta_t+1 is the next choice/value of theta (also add a brief reminder with this that alpha is the learning rate).

junseo-park replied:
Clarified in the "Brief Review" section what each of the variables means; it should be clearer what $t$ is, so I've left $t$ in for "This process is then repeated at each iteration $t$ of the gradient descent algorithm"

ananthagarwal commented on notebooks/ch11/gradient_stochastic.ipynb:
```diff
@@ -0,0 +1,285 @@
+    "\\end{align}\n",
+    "$$\n",
+    "\n",
+    "Stochastic gradient descent relies on the random selection of individual observations. This is statistically founded, since the randomness means that $E[\\nabla_{\\hat\\theta}l(\\hat\\theta, y_i)] = \\nabla_{\\hat\\theta}L(\\hat\\theta, y)$. In practice, we choose these random data points by shuffling the training data and iteratively selecting from the shuffled data. An iteration through th ...
```
(very) nit: Personally I think the comma in "statistically founded, since" should be removed to make it flow better but feel free to leave it in.

junseo-park replied:
Fixed!

ananthagarwal commented on notebooks/ch11/gradient_stochastic.ipynb:
```diff
@@ -0,0 +1,285 @@
+    "\\end{align}\n",
+    "$$\n",
+    "\n",
+    "Stochastic gradient descent relies on the random selection of individual observations. This is statistically founded, since the randomness means that $E[\\nabla_{\\hat\\theta}l(\\hat\\theta, y_i)] = \\nabla_{\\hat\\theta}L(\\hat\\theta, y)$. In practice, we choose these random data points by shuffling the training data and iteratively selecting from the shuffled data. An iteration through th ...
```
I think it would be helpful to clarify that the end of an epoch is when all n training data points have been used for the weight update.

Also personally I found the "sees examples from various classes" description to be a little confusing. Perhaps we can reword it as "ensure that the model sees the data points in an arbitrary order."

ananthagarwal commented on notebooks/ch11/gradient_stochastic.ipynb:
```diff
@@ -0,0 +1,285 @@
+    "![](gd.png)\n",
+    "![](sgd.png)\n",
+    "\n",
+    "At each iteration of batch gradient descent, we move in the direction of the true gradient of the cost function, which is depicted by the ellipses. On the other hand, each iteration of stochastic gradient descent may not lead us in the direction of the true gradient; however, the $\\hat\\theta$ parameters eventually converge to the minima of the cost function. Surprisingly, stochastic gradie ...
```
I remember a vitamin question that said "Takes less update steps to converge" is not a strength of SGD, so if you are referring to SGD computation time it would be good to clarify that this isn't necessarily due to fewer update steps.

ananthagarwal commented on notebooks/ch11/gradient_stochastic.ipynb:
```diff
@@ -0,0 +1,285 @@
+   "source": [
+    "## Mini-batch Gradient Descent\n",
+    "\n",
+    "**Mini-batch gradient descent** strikes a balance between batch gradient descent and stochastic gradient descent by increasing the number of observations that we select at each iteration. In mini-batch gradient descent, we take a simple random sample of observations called a mini-batch. We use the average of the gradients of their loss functions to construct an estimate of the true gradient  ...
```
Reminder to change "cost" to "loss"

junseo-park replied:
👍

ananthagarwal commented on notebooks/ch11/gradient_stochastic.ipynb:
```diff
@@ -0,0 +1,285 @@
+    "\n",
+    "We use batch gradient descent to iteratively improve model parameters until the model achieves low cost. Since batch gradient descent is computationally intractable with large datasets, we often use stochastic gradient descent to fit models instead. With a GPU, mini-batch gradient descent can converge more quickly than stochastic gradient descent.\n",
+    "\n",
+    "Although stochastic gradient descent typically takes more individual updates to converge than batch gradient descent, stochastic gradient descent often converges more quickly because it computes updates faster."
```
This last line "Although stochastic gradient descent ..." addresses my comment above about clarifying that SGD may take more individual steps but still converge faster. I think this sentence would be better placed after the two images, and you could just have a short line in the summary that SGD is computationally faster than batch GD.

junseo-park replied:
👍

ananthagarwal commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,187 @@
+
+## Gradient Descent in Logistic Regression
+
+Previously, we covered batch gradient descent, an algorithm that iteratively updates $\hat\theta$ to find the cost-minimizing parameters. We also discussed stochastic gradient descent and mini-batch gradient descent, methods that take advantage of statistical theory and parallelized hardware to decrease the time spent training the gradient descent algorithm. In this section, we will apply these c ...
```
(Sorry for making comments on the .ipynb file for your Ch 11 section, it's probably easier to understand to the comments on .md files so I'll switch to that here).

nit: Replace "Previously" with "In chapter 11"

junseo-park replied:
Sam said we're trying not to reference chapter numbers, since they might be changed around in future versions of the textbook (case in point: classification just got moved from ch16 to ch17)

ananthagarwal commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,187 @@
+\hat\theta_{t+1} = \hat\theta_t - \alpha \cdot \nabla_\hat\theta L(\hat\theta, X, y)
+$$
+
+In logistic regression, we use the cross entropy cost as our cost function. $\nabla_\hat\theta L(\hat\theta, X, y)$ is then the gradient of the cross entropy cost, $-\frac{1}{n}\sum_{i=1}^n(y_i - \sigma_i)X_i$; plugging this in allows us to find the gradient descent algorithm specific to logistic regression. Letting $ \sigma_i = f_\hat\theta(X_i) = \sigma(X_i \cdot \hat \theta) $, this becomes:
```
General reminder to replace "cost" with "loss" for this section.

I think it would be best to center the cross-entropy equation "L(theta, X, y) = 1/n sigma ..." and then write "(inverted delta_theta L(theta, X, y)  is then the gradient of the cross-entropy loss", mainly because we haven't usually put full equations as part of a text paragraph.

junseo-park replied:
Good suggestion; fixed.

ananthagarwal commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,187 @@
+
+### Stochastic Gradient Descent
+
+The general update formula is below, where $l(\hat\theta, X, y)$ is the loss function for a single data point:
```
I may be interpreting this wrong, but shouldn't it be X_i and y_i for a single data point instead of X and y?

junseo-park replied:
You're right, that's a typo!

ananthagarwal commented on notebooks/ch16/classification_sgd.md:
```diff
@@ -0,0 +1,187 @@
+
+## Implementation in Scikit-Learn
+
+Scikit-learn has implementations for stochastic gradient descent and mini-batch gradient descent using the [`SGDClassifier`](http://scikit-learn.org/stable/modules/generated/sklearn.linear_model.SGDClassifier.html) class. To gain a better estimate of the speed benefits afforded by stochastic gradient descent and mini-batch gradient descent, we will first manually implement batch gradient descent; ...
```
Include a short one-two line description of what the code boxes are doing, along the lines of "the code below trains a classifier predicting LeBron's shots using the three GD algorithms, and you can see that ____ is the fastest."

junseo-park commented:
I've added the following changes:
- Updated [10.2 (Modeling and Estimation/Cost Functions)](https://www.textbook.ds100.org/ch10/modeling_cost_functions.html) to use the same sample dataset as [11.2 (Gradient Descent/Defining Gradient Descent)](https://www.textbook.ds100.org/ch11/gradient_descent_define.html).
- Made fixes to all but one of Ananth's suggestions

The last of Ananth's suggestions involves the example generation that I've been having trouble with. The initial plan was to (1) include a manually-coded example in the Gradient Descent chapter using the tips dataset; (2) include Batch GD, SGD, and Mini-Batch GD examples in ch16 using sklearn's SGDClassifier, preferably using lebron.csv. For ch11: even after toying with alpha and the threshold values for SGD and Mini-Batch GD, I'm having trouble finding a combination that converges, let alone have a lower runtime than Batch GD. For ch16: Following Ananth's advice, I'll soon be trying out the spam emails dataset to see if that's a suitable example. As of this commit, these efforts remain unfinished in:
- 3 sets of code chunks in notebooks/ch11/gradient_stochastic.ipynb
- An unfinished blob at the end of notebooks/ch16/classification_sgd.ipynb

SamLau95 commented on notebooks/ch11/gradient_stochastic.md:
```diff
@@ -0,0 +1,331 @@
+
+
+
+This process is then repeated at each iteration $t$ of the gradient descent algorithm. For large $n$, this can become a computationally expensive problem to solve.
```
"This process" is vague. How about:

> Thus, for each iteration of batch gradient descent, we must compute the loss on every point in the dataset. Since datasets can contain many points, batch gradient descent can take a prohibitive amount of time to converge.

SamLau95 commented on notebooks/ch11/gradient_stochastic.md:
```diff
@@ -0,0 +1,331 @@
+
+## Stochastic Gradient Descent
+
+To circumvent the difficulty of computing a gradient across the entire training set, **stochastic gradient descent** approximates the overall gradient using a single randomly chosen data point. Since the observation is chosen randomly, we expect that using the gradient at each individual observation will eventually converge the algorithm to the same parameters as batch gradient descent. The gener ...
```
For the last sentence, how about:

> The modified update formula for stochastic gradient descent is presented below, where $l(\hat\theta, y_i)$ is the loss function for a single data point.

SamLau95 commented on notebooks/ch11/gradient_stochastic.md:
```diff
@@ -0,0 +1,331 @@
+\end{align}
+$$
+
+Stochastic gradient descent relies on the random selection of individual observations. This is statistically founded since the randomness means that $E[\nabla_{\hat\theta}l(\hat\theta, y_i)] = \nabla_{\hat\theta}L(\hat\theta, y)$. In practice, we choose these random data points by shuffling the training data and iteratively selecting from the shuffled data. An iteration through all $n$ observatio ...
```
We can't include the second sentence since we haven't introduced expectation at this point—let's just omit it.

"arbitrary order" and "random order" have very different meanings in statistics. I think we should just end the last sentence with "we reshuffle the data". (Technically the data are already in a random order after the first shuffle.)

SamLau95 commented on notebooks/ch11/gradient_stochastic.md:
```diff
@@ -0,0 +1,331 @@
+
+### Defining a Function for Stochastic Gradient Descent
+
+As we previously did for batch gradient descent, we would like to define a function that computes the stochastic gradient descent of the loss function. It will be similar to our `minimize` function, but we will need to implement the random selection of one observation at each iteration.
```
We don't need to say "would like" since we actually define it. Let's just keep this paragraph in present tense.

The last sentence shouldn't have a comma before the "but" since the first part of the sentence can't become a standalone sentence.

SamLau95 commented:
Oh, I forgot to add this yesterday, but you'll notice that many of my comments in your past reviews are about ambiguous subjects. I suggest being very careful when using pronouns and the word "this"—a bit of extra repetition is better than being unclear.
